File: LC200/lc164.js
function maximum_gap(nums) {
    if (nums.length < 2) {
        return 0;
    }
    var N = nums.length;
    var minimum = Math.min(...nums);
    var maximum = Math.max(...nums);
    if (minimum === maximum) {
        return 0;
    }
    var gap = parseInt(Math.ceil(parseFloat(maximum-minimum)/(N-1)));
    var num_buckets = parseInt(Math.ceil(parseFloat(maximum-minimum) / gap));
    var bucket_min = new Array(num_buckets).fill(null);
    var bucket_max = new Array(num_buckets).fill(null);
    var index, curMin, curMax;
    for (var number of nums) {
        if (number === minimum || number === maximum) {
            continue;
        }
        index = Math.floor((number - minimum) / gap);
        curMin = bucket_min[index];
        curMax = bucket_max[index];
        if (curMin === null) {
            bucket_min[index] = number;
            bucket_max[index] = number;
        } else {
            bucket_min[index] = Math.min(curMin, number);
            bucket_max[index] = Math.max(curMax, number);
        }
    }
    var max_gap = 0;
    var previous = minimum;
    for (var i=0; i<num_buckets; i++) {
        if (bucket_min[i] === null) {
            continue;
        }
        max_gap = Math.max(max_gap, bucket_min[i] - previous);
        previous = bucket_max[i];
    }
    return Math.max(max_gap, maximum - previous);
}

function test() {
    var nums_list = [
        [],
        [1, 1],
        [10, 21, 17, 13, 22, 23, 28, 30],
        [1, 1, 1, 1, 5, 5, 5, 5],
        [41, 42, 81],
        [2, 6]
    ];

    for (var nums of nums_list) {
        console.log(maximum_gap(nums));
    }
}

test();

File: LC200/lc164.py
"""
Given an unsorted array, find the maximum difference
 between the successive elements in its sorted form.

Try to solve it in linear time/space.

Return 0 if the array contains less than 2 elements.

You may assume all elements in the array are non-negative integers
 and fit in the 32-bit signed integer range.
"""

import math


def maximum_cap(nums):
    if len(nums) < 2:
        return 0

    length = len(nums)
    minimum = min(nums)
    maximum = max(nums)
    if minimum == maximum:
        return 0

    gap = int(math.ceil(float(maximum - minimum) / (length - 1)))
    num_buckets = int(math.ceil(float(maximum - minimum) / gap))
    bucket_min = [None] * num_buckets
    bucket_max = [None] * num_buckets

    for number in nums:
        if number == minimum or number == maximum:
            continue

        index = (number - minimum) // gap  # integer = integer / integer
        cur_min, cur_max = bucket_min[index], bucket_max[index]
        if cur_min is None:
            bucket_min[index] = number
            bucket_max[index] = number
        else:
            bucket_min[index] = min(number, cur_min)
            bucket_max[index] = max(number, cur_max)

    max_cap = 0
    previous_max = minimum
    for i in range(num_buckets):
        if bucket_min[i] is None:
            continue
        max_cap = max(max_cap, bucket_min[i] - previous_max)
        previous_max = bucket_max[i]

    return max(max_cap, maximum - previous_max)


def test():
    nums_list = [
        [],
        [1, 1],
        [10, 21, 17, 13, 22, 23, 28, 30],
        [1, 1, 1, 1, 5, 5, 5, 5],
        [41, 42, 81],
        [2, 6]
    ]

    for nums in nums_list:
        print(maximum_cap(nums))


test()
